include/rmm/device_buffer.hpp

```diff
@@ -236,6 +236,36 @@ class device_buffer {
     _stream = cuda_stream_view{};
   }
 
+  /**
+   * @brief Increase the capacity of the device memory allocation
+   *
+   * If the requested `new_capacity` is less than or equal to `capacity()`, no
+   * action is taken.
+   *
+   * If `new_capacity` is larger than `capacity()`, a new allocation is made on
+   * `stream` to satisfy `new_capacity`, and the contents of the old allocation are
+   * copied on `stream` to the new allocation. The old allocation is then freed.
+   * The bytes from `[size(), new_capacity)` are uninitialized.
+   *
+   * @throws rmm::bad_alloc If creating the new allocation fails
+   * @throws rmm::cuda_error if the copy from the old to new allocation
+   * fails
+   *
+   * @param new_capacity The requested new capacity, in bytes
+   * @param stream The stream to use for allocation and copy
+   */
+  void reserve(std::size_t new_capacity, cuda_stream_view stream)
+  {
+    set_stream(stream);
+    if (new_capacity > capacity()) {
+      auto tmp            = device_buffer{new_capacity, stream, _mr};
+      auto const old_size = size();
+      RMM_CUDA_TRY(cudaMemcpyAsync(tmp.data(), data(), size(), cudaMemcpyDefault, stream.value()));
+      *this = std::move(tmp);
```

Comment:
If you just move `tmp` over `*this`, is the old memory of `this` properly deallocated? I want to ensure there is no memory leak. With `swap`, it's obvious to me there is no leak because `tmp` is on the stack and when it goes out of scope it will be destroyed, taking the old memory of `this` with it.

Reply:
>  is the old memory of this properly deallocated?

Yep, it will invoke the move assignment operator of `device_buffer` which will deallocate the original memory.

https://github.com/rapidsai/rmm/blob/be9b9a96aecac44ffb03ee9681ce1cf17027be65/include/rmm/device_buffer.hpp#L194-L223

Reply:
I agree the move is safe. However I have a question about streams. Should the old buffer be destroyed on the same stream as the copy (to ensure the copy is complete), or the stream the old buffer was constructed with? Should the _copy_ always occur on the stream used to construct the original buffer to ensure that the reserve sequences after the constructor’s allocation? (I don’t remember seeing an explicit sync but need to look again.)

Reply:
> Is this solved by the call to set_stream?

Yes. Everything here is happening all on the provided `stream` argument. 

It would be an error to construct a `device_buffer` on `s1` and reserve on `s2` without first doing a synchronization between `s1` and `s2`.

Reply:
Fantastic. I wanted to be sure I understood that correctly. Approving now.

Reply:
Thanks all. Sounds right.